### PR TITLE
Enable dynamic filter pushdown for LEFT/RIGHT/SEMI/ANTI/Mark joins; surface probe metadata in plans; add join-preservation docs

### LINKED ISSUE #16973: Join DynamicFilter enhancements
### Tracking issue for improvements to dynamic filters w/ joins

- [ ] Generate filters for more join types
- [ ] Allow parent filters to pass through in more cases (https://github.com/apache/datafusion/pull/20192)
- [ ] Support nested joins (https://github.com/apache/datafusion/issues/20214)

#### Generate filters for more join types

In https://github.com/apache/datafusion/pull/16954 we are only going to support inner joins because of complexities with other join types.

This would benefit cases such as:

```sql
COPY (SELECT 1 AS c1) TO 't1.parquet';
COPY (SELECT 1 AS c1, 2 AS c2) TO 't2.parquet';

CREATE EXTERNAL TABLE t1 (
    c1 INT NOT NULL
)
STORED AS PARQUET
LOCATION  't1.parquet';

CREATE EXTERNAL TABLE t2 (
    c1 INT NOT NULL,
    c2 INT NOT NULL
)
STORED AS PARQUET
LOCATION  't2.parquet';

select *
from t1
left join t2 using (c1)
order by t2.c2
limit 100;
```

### Describe the solution you'd like

An implementation similar to that of the logical optimizer that accounts for the edge cases and allows some filters to be pushed down.

### PR DESCRIPTION
**Pull request by kosiew**
## Which issue does this PR close?

* Closes #16973

## Rationale for this change

Dynamic filter pushdown can significantly reduce I/O and compute by pruning non‑matching partitions/files on the fly. Previously this optimization was limited in scope. This PR expands it to additional join types and clarifies which input is pruned for each join. It also makes execution plans more transparent by explicitly annotating the probe side/keys for `HashJoinExec`, aiding debugging and user understanding.

## What changes are included in this PR?

* **Optimizer / Execution**

  * Enable dynamic filter pushdown for `LEFT`, `RIGHT`, `SEMI`, `ANTI`, and `LeftMark` joins (full joins remain unsupported; only equi-join keys contribute).
  * `HashJoinExec` now records and exposes `probe_side` and `probe_keys` in displayed physical plans to reflect runtime pruning behavior.
  * **Breaking:** `DynamicFilterPhysicalExpr::update` now requires an additional `key_count` argument to handle composite/equi key tracking.

* **Tests**

  * Update `sqllogictest` expected plans to include `probe_side=...` and `probe_keys=...` across tpch and other suites (e.g., `predicates.slt`, `subquery.slt`, `union.slt`, and multiple `tpch/plans/q*.slt.part`).

This PR updates 30 slt, slt.part files.

* **Documentation**

  * Add `docs/source/library-user-guide/join-preservation.md` summarizing which join sides are preserved for output filtering and ON‑clause evaluation.
  * Extend **Upgrading guide** with a new section describing dynamic filter pushdown for joins, configuration knobs, limitations (null handling, non‑equi predicates), and examples.
  * Update **User config docs** (`configs.md`) to clarify the semantics of `datafusion.optimizer.enable_dynamic_filter_pushdown`, including the dynamic filter target per join type and prerequisites for pushdown (e.g., Parquet pushdown enabled).
  * Draft **50.0.0 changelog** entry detailing the breaking API change and new dynamic filter pushdown behavior, with guidance and references.

## Are these changes tested?

Yes. This PR updates the `sqllogictest` goldens to reflect the new plan annotations and the broader dynamic filter pushdown coverage:

* Multiple updated `.slt` and `tpch` plan snapshots now show `probe_side`/`probe_keys` on `HashJoinExec` where applicable.
* Existing query result correctness is preserved; changes are confined to physical plan shape/metadata and runtime pruning behavior.
* Formats without predicate pushdown (CSV/JSON) continue to pass unchanged semantically; they simply do not benefit from pruning.

## Are there any user-facing changes?

* **Yes (docs/observability only):**

  * Physical plans printed in EXPLAIN now include `probe_side` and `probe_keys` for `HashJoinExec`.
  * New user‑guide page on join preservation and expanded configuration docs.
  * Changelog entry for 50.0.0.

* **Breaking API change:**

  * `DynamicFilterPhysicalExpr::update` requires a new `key_count` argument. Downstream users implementing custom dynamic filters must update their code.
  * Please add the `api change` label.

* **Behavioral notes (non‑breaking):**

  * Dynamic filter pushdown across more join types is **enabled by default** and can be disabled via `datafusion.optimizer.enable_dynamic_filter_pushdown=false`.
  * Effectiveness depends on source format pushdown support (e.g., Parquet with `execution.parquet.pushdown_filters=true`).
  * Only equi‑join conjuncts generate dynamic filters; full joins and non‑equi predicates remain out of scope for now.

### CODE REVIEW

File: datafusion/optimizer/src/push_down_filter.rs [hunk 1]
```diff
@@ -2730,7 +2787,7 @@ mod tests {
         assert_optimized_plan_equal!(
             plan,
             @r"
-        Right Join: Using test.a = test2.a
+        Right Join: Using test.a = test2.a Filter: test.a <= Int64(1)
```
Comment (Dandandan) [inline datafusion/optimizer/src/push_down_filter.rs]: These filters seem redundant (always true)?

**Comment by adriangb** [comment]
Does this actually close #16973? I don't see that it changes filter pushdown at the physical plan level at all which is what #16973 is talking about. The example in that issue involves a TopK and HashJoin operator.

**Comment by kosiew** [comment]
@Dandandan , @adriangb ,

This PR is not ready for review yet.

**Comment by adriangb** [comment]
> @Dandandan , @adriangb ,
> 
> 
> 
> This PR is not ready for review yet.

Yes but my point is that unless it changes drastically I don't think it addresses the issue it purports to close. This PR deals with logical expressions / plans but the issue is in reference to physical expressions / plans. Do you plan to expand the scope of this PR, or should we unlink the issue?

**Comment by kosiew** [comment]
@adriangb 
PR is ready for review.

**Comment by adriangb** [comment]
This is a monumental piece of work, I’m astounded! Thank you so much for working on this.

I’ll try to review it but I immediately will ask if we can somehow split it up into multiple logical PRs to make it easier to review and isolate future issues that may pop up.

I’m also wondering if you’ve seen https://github.com/apache/datafusion/issues/17196 and https://github.com/apache/datafusion/issues/17188. I imagine this PR may fix them but it would be nice to fix those bugs first with a  more targeted change / smaller PR first before adding more complexity.

**Comment by adriangb** [comment]
I also think it's important that we get something like #17177 in place before adding more support to avoid regressions

File: datafusion/common/src/join_type.rs [hunk 2]
```diff
@@ -111,6 +111,63 @@ impl JoinType {
                 | JoinType::RightAnti
         )
     }
+    /// Returns true if the left side of this join preserves its input rows
+    /// for filters applied *after* the join.
+    #[inline]
```
Comment (adriangb) [inline datafusion/common/src/join_type.rs:116]: I see a lot of use of `#[inline]`. My understanding is that without specific evidence that it helps performance it may actually sometimes hurt it and it's best to not throw it around unless it's very obvious or can be proven to help performance.

File: datafusion/common/src/join_type.rs [hunk 3]
```diff
@@ -111,6 +111,63 @@ impl JoinType {
...
+    pub const fn preserves_left_for_output_filters(self) -> bool {
```
Comment (adriangb) [inline datafusion/common/src/join_type.rs:117]: I arrived at this same refactor in https://github.com/apache/datafusion/pull/17153 - I think it's a good one. Can we pull this out into its own PR?

File: datafusion/core/tests/dataframe/mod.rs [hunk 4]
```diff
@@ -2664,23 +2664,23 @@ async fn test_count_wildcard_on_where_in() -> Result<()> {
...
+    |               |   SubqueryAlias: __correlated_sq_1                                                                                                                    |
+    |               |     Projection: count(Int64(1)) AS count(*)                                                                                                           |
+    |               |       Aggregate: groupBy=[[]], aggr=[[count(Int64(1))]]                                                                                               |
+    |               |         TableScan: t2 projection=[]                                                                                                                   |
+    | physical_plan | CoalesceBatchesExec: target_batch_size=8192                                                                                                           |
+    |               |   HashJoinExec: mode=CollectLeft, join_type=RightSemi, on=[(count(*)@0, CAST(t1.a AS Int64)@2)], projection=[a@0, b@1], probe_side=Left, probe_keys=0 |
```
Comment (adriangb) [inline datafusion/core/tests/dataframe/mod.rs:2677]: Can we pull out the addition of these fields to the debug output into its own PR?

File: datafusion/physical-expr/src/expressions/dynamic_filters.rs [hunk 5]
```diff
@@ -48,6 +51,9 @@ pub struct DynamicFilterPhysicalExpr {
     remapped_children: Option<Vec<Arc<dyn PhysicalExpr>>>,
     /// The source of dynamic filters.
     inner: Arc<RwLock<Inner>>,
+    /// Number of keys currently contained in this dynamic filter.
+    /// Uses relaxed atomics as this counter is for diagnostics only.
+    key_count: Arc<AtomicUsize>,
```
Comment (adriangb) [inline datafusion/physical-expr/src/expressions/dynamic_filters.rs:56]: If I'm just reading this module I'd have no idea what `key_count` is. Is it a join specific thing? What is a "key" in this context? If we actually do need this I propose we add it in an isolated PR.

File: datafusion/physical-expr/src/expressions/dynamic_filters.rs [hunk 6]
```diff
@@ -217,8 +234,25 @@ impl DynamicFilterPhysicalExpr {
         current.expr = new_expr;
         // Increment the generation to indicate that the expression has changed.
         current.generation += 1;
+        // Relaxed ordering is sufficient as `key_count` is only used for
+        // observability and does not synchronize with other data.
```
Comment (adriangb) [inline datafusion/physical-expr/src/expressions/dynamic_filters.rs:238]: observability is super vague - what is it actually used for?

File: datafusion/physical-optimizer/src/filter_pushdown.rs [hunk 7]
```diff
@@ -708,6 +711,7 @@ impl<T: Clone> FilteredVec<T> {
     }
 }
 
+#[inline]
```
Comment (adriangb) [inline datafusion/physical-optimizer/src/filter_pushdown.rs:714]: Same comment about putting inlines without any obvious justification

File: docs/source/library-user-guide/join-preservation.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  under the License.
+-->
+
+# Join preservation
+
+Dynamic filter pushdown and other optimizations rely on whether a join preserves
+rows from its inputs. The tables below summarise which sides are preserved for
+post-join output filtering and for evaluation of `ON`-clause predicates.
+
+## Output filtering
```
Comment (adriangb) [inline docs/source/library-user-guide/join-preservation.md:26]: Thanks so much. This is an instant commit / approve (with verification that it's correct) if made as it's own PR.

File: dev/changelog/50.0.0.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+- **Breaking:** `DynamicFilterPhysicalExpr::update` now requires an extra
+  `key_count` argument.
+- Enable dynamic filter pushdown for left, right, semi, anti, and mark joins
+  [#16445](https://github.com/apache/datafusion/pull/16445) (adriangb). Mark joins
+  push filters to the side opposite the preserved input (`dynamic_filter_side`; see
+  [tests](https://github.com/apache/datafusion/blob/main/datafusion/physical-plan/src/joins/hash_join.rs#L2033-L2049)). Formats without predicate pushdown (CSV/JSON) will not benefit.
+  Full joins and non‑equi (range or composite) predicates are not yet supported;
+  see [#7955](https://github.com/apache/datafusion/issues/7955). Dynamic filters
+  add planning overhead for high-cardinality keys; disable via:
```
Comment (adriangb) [inline dev/changelog/50.0.0.md:32]: Could you help me understand where the overhead comes from at the moment? Is it measurable at all?

File: dev/changelog/50.0.0.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+# Apache DataFusion 50.0.0 Changelog
+
+This release is under active development.
+
+- **Breaking:** `DynamicFilterPhysicalExpr::update` now requires an extra
+  `key_count` argument.
+- Enable dynamic filter pushdown for left, right, semi, anti, and mark joins
+  [#16445](https://github.com/apache/datafusion/pull/16445) (adriangb). Mark joins
+  push filters to the side opposite the preserved input (`dynamic_filter_side`; see
+  [tests](https://github.com/apache/datafusion/blob/main/datafusion/physical-plan/src/joins/hash_join.rs#L2033-L2049)). Formats without predicate pushdown (CSV/JSON) will not benefit.
```
Comment (adriangb) [inline dev/changelog/50.0.0.md:29]: Actually if they have statistics they will still benefit (those two generally don't)

File: datafusion/physical-plan/src/topk/mod.rs [hunk 11]
```diff
@@ -404,9 +404,11 @@ impl TopK {
             .reduce(|a, b| Arc::new(BinaryExpr::new(a, Operator::Or, b)));
 
         if let Some(predicate) = dynamic_predicate {
-            if !predicate.eq(&lit(true)) {
-                filter.update(predicate)?;
-            }
+            filter.update(predicate, self.heap.len())?;
+        } else {
+            // Even when the dynamic predicate is a tautology we still update
```
Comment (adriangb) [inline datafusion/physical-plan/src/topk/mod.rs:409]: tautology is a very fancy word - would it be possible to give an explanation of what the term means?

**Comment by alamb** [comment]
> Amazing work overall! A lot of the diff is updating debug outputs / slt tests. I think it will help a lot to split this up into multiple PRs so that e.g. that can be reviewed separately from a smaller diff with complex logic changes.

I agree with @adriangb that splitting this into multiple PRs would make this much easier to thoroughly review

From my past experience, handling pushdown for outerjoins in general is quite subtle (especially in the presence of nulls, etc) and I have chased down many very subtle bugs in my past lives. Testing (with nulls) is especially important too

**Comment by adriangb** [comment]
> Testing (with nulls) is especially important too

Yes a LOT of specific tests + fuzz tests are going to be needed to be certain we don't introduce bugs

**Comment by kosiew** [comment]
Thanks @adriangb , @alamb for your review.

Putting this into draft and I will split this into smaller PRs.

**Comment by adriangb** [comment]
> Thanks @adriangb , @alamb for your review.
> 
> Putting this into draft and I will split this into smaller PRs.

Amazing, thank you! Also please consider working on the fuzz tests - they'll help make sure we don't create bugs.

**Comment by kosiew** [comment]
@adriangb ,
> please consider working on the fuzz tests - they'll help make sure we don't create bugs.

Do you mean #17177 uncovered some edge cases that Dynamic Filter push down should address?

**Comment by adriangb** [comment]
> @adriangb ,
> 
> > please consider working on the fuzz tests - they'll help make sure we don't create bugs.
> 
> 
> 
> Do you mean #17177 uncovered some edge cases that Dynamic Filter push down should address?

No I'm just saying we do need to merge that or similar before making more changes

**Comment by adriangb** [comment]
@kosiew is this something your still actively working on? It's very cool work and I would love to help get it across the line. I'm going to work on the fuzz tests next week. Are you able to help make a plan to split this PR up into smaller units of work?

**Comment by kosiew** [comment]
hi @adriangb,

Yep, I can start on breaking this into smaller PRs this week.

**Comment by adriangb** [comment]
Amazing!
